grpc-protocol/src/main/java/com/linecorp/armeria/client/grpc/protocol/UnaryGrpcClient.java

```diff
@@ -159,32 +162,48 @@ public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) {
 
-                           try (ArmeriaMessageDeframer deframer = new ArmeriaMessageDeframer(new Listener() {
-                               @Override
-                               public void messageRead(DeframedMessage unframed) {
-                                   final ByteBuf buf = unframed.buf();
-                                   // Compression not supported.
-                                   assert buf != null;
-                                   responseFuture.complete(HttpResponse.of(
-                                           msg.headers(),
-                                           HttpData.wrap(buf).withEndOfStream(),
-                                           msg.trailers()));
-                               }
-
-                               @Override
-                               public void endOfStream() {
-                                   if (!responseFuture.isDone()) {
-                                       responseFuture.complete(HttpResponse.of(msg.headers(),
-                                                                               HttpData.empty(),
-                                                                               msg.trailers()));
-                                   }
-                               }
-                           }, Integer.MAX_VALUE, ctx.alloc(), false)) {
-                               deframer.request(1);
-                               deframer.deframe(msg.content(), true);
-                           }
+                           final ArmeriaMessageDeframerHandler handler =
+                                   new ArmeriaMessageDeframerHandler(Integer.MAX_VALUE);
+                           final HttpDeframer<DeframedMessage> deframer =
+                                   new HttpDeframer<>(handler, ctx.alloc());
+
+                           StreamMessage.of(msg.content()).subscribe(deframer, ctx.eventLoop());
+                           deframer.subscribe(singleSubscriber(msg, responseFuture), ctx.eventLoop());
                            return responseFuture;
                        }), ctx.eventLoop());
         }
+
+        private static Subscriber<DeframedMessage> singleSubscriber(
+                AggregatedHttpResponse msg, CompletableFuture<HttpResponse> responseFuture) {
+
+            return new Subscriber<DeframedMessage>() {
+                @Override
+                public void onSubscribe(Subscription s) {
+                    s.request(1);
+                }
+
+                @Override
+                public void onNext(DeframedMessage unframed) {
+                    final ByteBuf buf = unframed.buf();
+                    // Compression not supported.
+                    assert buf != null;
+                    responseFuture.complete(HttpResponse.of(msg.headers(), HttpData.wrap(buf).withEndOfStream(),
+                                                            msg.trailers()));
```

Comment:
If we set `withEndOfStream` to the body, trailers will not be sent.

Reply:
I see. Let me add a test case and fix it.

Reply:
This is an HttpResponse so I guess trailes could be delivered to a client even if `withEndOfStream` is set. However, I agree that it is better to remove `withEndOfStream` to avoid an additional allocation.

Reply:
https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java#L213
Doesn't this line call `subscription.cancel()` so the trailers is not sent?

Reply:
Ah, this is the client-side. 😅